Cloud statistics (#66)
eurec4a/how_to_eurec4a: how_to_eurec4a/cloud_stat_tools.md

One approach to this problem is the [group - apply - combine](https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html) approach, which is what among others, `xarray`, `pandas` and SQL provide.
The idea is group our data into common sections (those would be one group per individual cloud) and then run a function (which reduces some data to a single value) on each group individually.
The resulting values would then be combined back to our final result.
While this is a neat way to think about the problem, it has a major drawback when applied to `xarray` or `pandas`: it is really slow.
The method is just a sneaky way of running a for loop over all clouds and thus deliberately avoids the use of `numpy` to speed things up.

But we can still look at the structure of such an operation and then think about how we could do something similar with better performance.

> So first, we need some way of identifying each individual cloud (let's call them `section_id`).

Author: I think my plan was to differentiate between sections (may be cloudy or not cloudy) and (cloud-) segments (which are only the clouds). I've to admit that this is not very clear at this point.

> So first, we need some way of identifying each individual cloud (let's call them `section_id`).

Author: I've reformulated this paragraph to clarify the use of `section_id`.

If we can then find a way to accumulate data into individual bins based on the `section_id`, we should be ready to compute further statistics.

```{code-cell} ipython3
section_ids = np.cumsum(np.abs(edges[:-1]))
```

```{code-cell} ipython3
:tags: [hide-input]

fig, (ax1, ax2, ax3) = plt.subplots(3, sharex=True)
ax1.step(np.arange(len(cloudmask)) + .5, cloudmask, where='mid')
ax1.spines["bottom"].set_visible(False)
ax1.set_aspect(.5)
ax1.set_title("cloudmask")

ax2.step(np.arange(len(edges)), edges, where='mid')
ax2.spines["bottom"].set_visible(False)
ax2.set_aspect(.5)
ax2.set_title("edges")

ax3.step(np.arange(len(section_ids)) + .5, section_ids, where='mid')
ax3.spines["bottom"].set_visible(False)
#ax3.set_aspect(.5)
ax3.set_title("section ids")
ax3.set_xlabel("locs")
ax3.set_xlim(0,20)
ax3.yaxis.set_major_locator(ticker.IndexLocator(base=2, offset=1))
ax3.xaxis.set_major_locator(ticker.MultipleLocator(1));
```

Let's have a look at the whole process idea in a graphical way:
```{figure} figures/cloud_segment_stats.png
:alt: computing cloud segment statistics
:width: 100%
:align: center
```

+++

> Let's say we want to compute the average optical thickness of each detected cloud segment.

Reviewer: ```suggestion
Let's say we want to compute the average optical thickness `ot` of each detected cloud segment.
```

Author: I think `ot` is currently used for the per-pixel optical thickness, not the average optical thickness. Do you mean that it would be better to change the per-pixel optical thickness to something else and use `ot` for the average?

For this example, we assume the following optical thickness per source pixel:
```{code-cell} ipython3
ot = np.array([0., 0., 1., 2., 4., 5., 0., 0., 0., 2., 2., 0., 0., 0., 0., 1., 1., 1., 0., 0.])
```
We already know how many pixels are part of the cloud (the cloud length from above), so we only need the sum of the optical thickness within the segment in order to compute the mean.
Thus, what we conceptually want to do is:
```python
ot_sum[section_ids] += ot
```
But `+=` will only work correctly if none of the index values from `section_ids` are duplicated, which definitely is not the case, as we designed `section_ids` such that it will contain the same value for every pixel within the same cloud.
Fortunately `numpy` provides [`ufunc.at`](https://numpy.org/doc/stable/reference/generated/numpy.ufunc.at.html) which solves this exact problem.

> The `at` method is available for all scalar universal functions of numpy which take two arguments and is intended for repeated in-place application of the given universal function.

Reviewer: ```suggestion
The `at` method is available for all scalar universal functions of `numpy` which take two arguments and is intended for repeated in-place application of the given universal function.
```

Addition (`add`) is such a scalar ufunc with two arguments, so we can use `np.add.at` in this case:

```{code-cell} ipython3
ot_sum = np.zeros(np.max(section_ids) + 1)
np.add.at(ot_sum, section_ids, ot)
ot_sum
```

````{note}
Observe how
```python
np.add.at(ot_sum, section_ids, ot)
```
can be used as a direct replacement for
```python
ot_sum[section_ids] += ot
```

> However, while the latter is simpler to understand intuitively, only `np.add.at` will provide the correct results.

Reviewer: ```suggestion
However, though the latter is simpler to understand, only `np.add.at` will provide the correct results.
```

````

As indicated above, the result will contain values for each clear and cloudy segment.

> In order to compute the final mean cloud optical thickness, we simply can use every odd element of `ot_sum` (those are the cloudy pieces) and divide them by cloud length:

Reviewer: Maybe I overlooked something, but isn't it only the case in this explicit example that "every odd element" is cloudy? What if your array starts with a cloud? If so, is it possible to formulate it in a more generic way?

Author: It's actually already formulated in a generic way, but maybe that's still a bit too hidden 😁...
In case the first pixel is a cloud, the first edge (before the first pixel) will be a starting edge (E_0 = C_0) and thus the very first `segment_id` will be 1 and thus odd. Probably that's worth a separate note though...

Author: I've added a note which should clarify this issue.
